File: backend/app/db/models/partners.py
from __future__ import annotations

from decimal import Decimal
from uuid import UUID

from sqlalchemy import (
    JSON,
    Boolean,
    CheckConstraint,
    ForeignKey,
    Numeric,
    String,
    UniqueConstraint,
    Uuid,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base
from app.db.models.mixins import IdentifiedTimestampMixin, OrganizationScopeMixin


class BusinessPartner(IdentifiedTimestampMixin, OrganizationScopeMixin, Base):
    __tablename__ = "business_partners"
    __table_args__ = (
        CheckConstraint(
            "partner_type IN ('customer', 'vendor', 'both')",
            name="ck_business_partners_type",
        ),
        CheckConstraint("credit_limit >= 0", name="ck_business_partners_credit_limit"),
        CheckConstraint("payment_terms_days >= 0", name="ck_business_partners_payment_terms"),
        UniqueConstraint(
            "organization_id", "partner_code", name="uq_business_partners_org_code"
        ),
    )

    partner_code: Mapped[str] = mapped_column(String(64), nullable=False)
    partner_type: Mapped[str] = mapped_column(String(20), nullable=False)
    display_name: Mapped[str] = mapped_column(String(250), nullable=False)
    legal_name: Mapped[str | None] = mapped_column(String(250), nullable=True)
    tax_id: Mapped[str | None] = mapped_column(String(32), nullable=True)
    tax_branch: Mapped[str] = mapped_column(String(20), default="00000", nullable=False)
    email: Mapped[str | None] = mapped_column(String(320), nullable=True)
    phone: Mapped[str | None] = mapped_column(String(50), nullable=True)
    payment_terms_days: Mapped[int] = mapped_column(default=0, nullable=False)
    credit_limit: Mapped[Decimal] = mapped_column(
        Numeric(18, 2), default=Decimal("0.00"), server_default="0", nullable=False
    )
    is_active: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    version: Mapped[int] = mapped_column(default=1, server_default="1", nullable=False)
    tags: Mapped[list[str]] = mapped_column(JSON, default=list, nullable=False)

    addresses: Mapped[list[PartnerAddress]] = relationship(
        back_populates="partner",
        cascade="all, delete-orphan",
        order_by="PartnerAddress.is_primary.desc()",
    )


class PartnerAddress(IdentifiedTimestampMixin, OrganizationScopeMixin, Base):
    __tablename__ = "partner_addresses"
    __table_args__ = (
        CheckConstraint(
            "address_type IN ('registered', 'billing', 'shipping', 'other')",
            name="ck_partner_addresses_type",
        ),
    )

    partner_id: Mapped[UUID] = mapped_column(
        Uuid(as_uuid=True), ForeignKey("business_partners.id", ondelete="RESTRICT"), nullable=False
    )
    address_type: Mapped[str] = mapped_column(String(20), default="other", nullable=False)
    label: Mapped[str | None] = mapped_column(String(100), nullable=True)
    address_line1: Mapped[str] = mapped_column(String(250), nullable=False)
    address_line2: Mapped[str | None] = mapped_column(String(250), nullable=True)
    district: Mapped[str | None] = mapped_column(String(120), nullable=True)
    province: Mapped[str | None] = mapped_column(String(120), nullable=True)
    postal_code: Mapped[str | None] = mapped_column(String(20), nullable=True)
    country_code: Mapped[str] = mapped_column(String(2), default="TH", nullable=False)
    is_primary: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)

    partner: Mapped[BusinessPartner] = relationship(back_populates="addresses")


__all__ = ["BusinessPartner", "PartnerAddress"]

File: frontend/lib/partners.ts
import { apiRequest } from "@/lib/api-client";

export type PartnerType = "customer" | "vendor" | "both";

export type PartnerAddress = {
  id: string;
  partner_id: string;
  address_type: "registered" | "billing" | "shipping" | "other";
  label: string | null;
  address_line1: string;
  address_line2: string | null;
  district: string | null;
  province: string | null;
  postal_code: string | null;
  country_code: string;
  is_primary: boolean;
};

export type Partner = {
  id: string;
  tenant_id: string;
  organization_id: string;
  partner_code: string;
  partner_type: PartnerType;
  display_name: string;
  legal_name: string | null;
  tax_id: string | null;
  tax_branch: string;
  email: string | null;
  phone: string | null;
  payment_terms_days: number;
  credit_limit: string;
  is_active: boolean;
  version: number;
  tags: string[];
  addresses: PartnerAddress[];
};

export type PartnerListFilters = {
  partnerType?: PartnerType;
  search?: string;
  includeArchived?: boolean;
};

export type CreatePartnerInput = {
  partner_code: string;
  partner_type: PartnerType;
  display_name: string;
  tax_id?: string;
  tax_branch?: string;
  email?: string;
  phone?: string;
  payment_terms_days?: number;
};

export function listPartners(
  filters: PartnerListFilters = {},
  signal?: AbortSignal,
): Promise<Partner[]> {
  const query = new URLSearchParams();
  if (filters.partnerType) {
    query.set("type", filters.partnerType);
  }
  if (filters.search?.trim()) {
    query.set("search", filters.search.trim());
  }
  if (filters.includeArchived) {
    query.set("include_archived", "true");
  }
  const suffix = query.toString() ? `?${query.toString()}` : "";
  return apiRequest<Partner[]>(`/v1/partners${suffix}`, { signal });
}

export function createPartner(
  input: CreatePartnerInput,
  idempotencyKey: string,
): Promise<Partner> {
  return apiRequest<Partner>("/v1/partners", {
    method: "POST",
    headers: { "Idempotency-Key": idempotencyKey },
    body: JSON.stringify(input),
  });
}

export function createPartnerIdempotencyKey(): string {
  let randomId = globalThis.crypto?.randomUUID?.();
  if (!randomId && globalThis.crypto?.getRandomValues) {
    const bytes = globalThis.crypto.getRandomValues(new Uint8Array(16));
    randomId = Array.from(bytes, (byte) => byte.toString(16).padStart(2, "0")).join("");
  }
  if (!randomId) {
    throw new Error("Secure randomness is unavailable");
  }
  return `partner-create-${randomId}`;
}
